test/fdk-test-kit/README.md

```diff
@@ -0,0 +1,77 @@
+Testing FDK-based functions
+===========================
+
+Function development kit (FDK) as a piece of software that helps to write hot functions by encapsulating logic of processing incoming requests with respect to defined protocol (HTTP, JSON).
+Particular testing framework help developers to identify if any programming language-specific FDK compatible with Fn hot formats.
+
+
+Prerequisites
+-------------
+
+This testing framework allows to run FDK tests against live Fn service, to let tests know of where Fn service is hosted please set following environment variable:
+```bash
+    export API_URL=http://fn.io:8080
+```
+
+As an alternative test suite capable to bootstrap its own copy of Fn service locally, for this particular case following environment variables must be set:
+```bash
+    export DOCKER_HOST=/var/run/docker.sock
+```
+or wherever Docker Remote API daemon listens.
+
+Test suite requires general purpose programming language-specific FDK-based function image that must be developed specifically for this test suite, following environment variable must be set:
+```bash
+    export FDK_FUNCTION_IMAGE="..."
+```
+This environment variable should contain a reference to the particular docker image.
+
+
+Test suite details
+------------------
+
+Test suite contains following tests:
+
+1. `TestFDKFormatSmallBody`
+
+`TestFDKFormatSmallBody` test
+--------------------------
+
+FDK should support following formats:
+
+ - HTTP (subtest: `test-fdk-http-small-body`)
```

Comment:
As long as Fn gives protocol options all FDKs must be compatible.

Reply:
agree w/ travis if we can have a nice way for tests just to say 'doesn't work with http' then a lot of fdks may be okay with that (we could still consider them 'compliant'?). it would be nice if images/code could expose their capabilities so that coordination wasn't needed between function configuration of format and the code itself, not sure how this would look, and somewhat unrelated to this patch.

Reply:
That's the point. We don't have criteria for compliance. If FDK supports only default format (which is not the case, because you don't need an FDK to work with default format) would that make FDK compliant? If FDK supports only one format distinct from the default (which is the case of java, go, ruby FDKs) does it make FDK compliant?

From my point of view, I do want tests to say that FDK is compatible/compliant if it implements all supported formats.

If we go the road of "compliant if supports at least one hot protocol" then we need to define the strategy for FDKs how to reject unsupported protocols, for instance: HTTP 501 Not Implemented (Unsupported protocol). This response should be mandatory in order to make tests keep going on because not implemented protocol is not an error in FDK internal parts rather than the choice of developer.

cc @zootalures

Reply:
> We don't have criteria for compliance.

i have some ideas about compliance if you want to start somewhere to collab on that, I'll poop the ideas out here:

* runs > 1 call in same image [hot], serially
* [eventually] times out a call properly (i.e. takes input of next call when expected)
* parses headers for the set of expected headers, e.g. can output correct CALL_ID on N runs
* certain vars accessible from env (make sure no masking of env)
* supports default and >= 1 hot format based on `FN_FORMAT` env var changing
* ideally, the code looks a certain way, but this requires humans or so-much-static-analysis-it's-not-worth-it in 2017. (i.e. want to make sure they have a 'context' variable that has config, headers on it)

maybe missing something, but overall they're pretty simple and that seems like a good start at least. if at the end we make sure >=1 hot format works and then thumbs up, that seems doable?

> then we need to define the strategy for FDKs how to reject unsupported protocols,

+1. it would be kinda sweet if they could back channel which formats _would_ work and fn could just switch to that (it seems possible, at least, but unrelated to this work).